emma-spark/src/main/scala/org/emmalanguage/api/SparkDataset.scala

```diff
@@ -116,6 +117,16 @@ class SparkDataset[A: Meta] private[api](@transient private[emmalanguage] val re
     rep.collect()
 
   // -----------------------------------------------------
+  // Pre-defined folds
+  // -----------------------------------------------------
+
+  override def find(p: A => Boolean): Option[A] =
+    Option(fold[Tuple1[A]](Tuple1(null.asInstanceOf[A]))(
+      x => if (p(x)) Tuple1(x) else Tuple1(null.asInstanceOf[A]),
+      (x, y) => if (x._1 != null) x else y
+    )._1)
```

Comment:
I envision more problems with this:
1. Any other fold based on `reduceOption` won't work on Spark Datasets, including `min` and `max`.
2. For primitive types such as `Int`, `x._1 != null` will always be true, thus returning the default value (e.g. `0`) instead of the one satisfying `p`. Note that `Tuple1` is `@specialized(Int, Long, Double)`.

A possible solution would be to use `(Boolean, A)` where the boolean encodes whether `A` is `Some` or `None`.